doc/lumaenc.1
.TH LUMAENC 1
.SH NAME
lumaenc \- Compress a sequence of high dyncamic range (HDR) frames in to a Matroska (.mkv) HDR video
.SH SYNOPSIS
.B lumaenc
\fB\-\-output \fIFILE\fR
[\fB\-\-input \fIFILE\fR]
[\fB\-\-start-frame \fIFRAME\fR]
[\fB\-\-end-frame \fIFRAME\fR]
[\fIOPTIONS\fR ...]
.SH DESCRIPTION
.B lumaenc
The application encodes a high dynamic range (HDR) video from a set of separate
HDR frames. The images are transformed with a perceptual transfer function (PTF)
to make the quantization error perceptually equally distributed over the different
luminances. The transformed frames are then encoded using Google's VP9 video codec,
using a high bit depth profile. Finally, the video is packaged together with all
needed meta data in a Matroska containter.

The parameters for \fBlumaenc\fR, listed below, provide a set of options for tweaking
the bit depth of the encoded video stream:

.RS
\fB--profile\fR           - selects encoding profile
.RE
.RS
\fB--ptf-bitdepth\fR      - selects bit depth target of perceptual transfer function
.RE
.RS 
\fB--color-bitdepth\fR    - selects bit depth target of color channels
.RE
.RS 
\fB--encoding-bitdepth\fR - selects the actual bit depth to encode at
.RE

The ptf and color bit depths are only the ranges to which the luminance and color 
channels are mapped to, respectively. They are then encoded using the selected encoding 
bit depth. This means that if the ptf or color bit depth is larger than the encoding 
bit depth, the values that do not fit within this bit depth will be clamped. And 
if the ptf or color bit depth is less than the encoding bit depth, there will be 
empty bits and thus an effective bit depth that is lower than the encoding bit 
depth (this is the default for luminance and color channels, which only use 11 and
8 bits, respectively, but are encoded at 12 bits).

This also means that some combinations of the options can not be used, and the encoder 
should correct for this. For example, if profile 0 is selected, then the encoding 
bit depth must be 8, since this profile only supports 8 bits. The encoder will detect
if the encoding bit depth is larger, and in that case change the profile to 2.


.SH OPTIONS
.TP
.B \-i  \fIFILE\fR, \fB\-\-input \fIFILE
Input HDR video sequence. The frames should either be in EXR format, or piped to
\fBlumaenc\fR from pfstools in which case this option should not be used. A range of frames
is specified using a %d pattern togheter with the \fB--frames\fR option.

.TP
.B \-o  \fIFILE\fR, \fB\-\-output \fIFILE
Output location of the compressed HDR video. The video is stored using the Matroska
container, and should thus use file extension '.mkv'.

.TP
.B \-f  \fIFRAMES\fR, \fB\-\-frames \fIFRAMES
Input frame range, formatted as startframe:step:endframe

.TP
.B \-fps  \fIVALUE\fR, \fB\-\-framerate \fIVALUE
Framerate of encoded HDR video stream, specified as frames/s.

Default value is 25.

.TP
.B \-p  \fIPROFILE\fR, \fB\-\-profile \fIPROFILE
VP9 encoding profile:

  0 : 8 bits,  422 color sub sampling
  1 : 8 bits,  444
  2 : 12 bits, 422 color sub sampling
  3 : 12 bits, 444
  
Default value is 2.

.TP
.B \-pb  \fIDEPTH\fR, \fB\-\-ptf-bitdepth \fIDEPTH
Bit depth of the perceptual transfer function. This should be <= \fB--encoding-bitdepth\fR
to fit the bit depth used for encoding.

Default value is 11.

.TP
.B \-cb  \fIDEPTH\fR, \fB\-\-color-bitdepth \fIDEPTH
Bit depth of the color channels. This should be <= \fB--encoding-bitdepth\fR
to fit the bit depth used for encoding.

Default value is 8.

.TP
.B \-eb  \fIDEPTH\fR, \fB\-\-encoding-bitdepth \fIDEPTH
Encode at 8 bits, or use high bit depth encoding at 10 or 12 bits.

Default is 12.

.TP
.B \-q  \fISCALE\fR, \fB\-\-quantizer-scaling \fISCALE
Scaling of the encoding quantization. This provides a direct control over the quality
of the encoded video, where 0 is the best quality and 63 is the lowest quality.

Default value is 2.

.TP
.B \-sc  \fISCALE\fR, \fB\-\-pre-scaling \fISCALE
Scaling of pixels to apply before transformation and encoding. Since the perceptual
encoding expects input to be in physical units (cd/m²), this option can be used
to calibrate the input to the correct range.

Default value is 1.

.TP
.B \-ptf  \fIPTF\fR, \fB\-\-transfer-function \fIPTF
The perceptual transfer function used for encoding:

  PSI     :  Perceptual - Ferwerda's t.v.i.
  PQ      :  Dolby perceptual quantizer (SMPTE ST 2084)
  LOG     :  Logarithmic transformation
  HDRVDP  :  Transformation function from HDR-VDP
  LINEAR  :  Linear scaling

Default is PQ.

.TP
.B \-ma  \fIMAXLUMINANCE\fR, \fB\-\-max-luminance \fIMAXLUMINANCE
Maximum luminance (cd/m2) for the perceptual transfer function. Values above 
this will be clipped. This option only applies to the PQ and LOG transfer 
functions, as the other available PTFs use lookup tables.

Default value is 10000.

.TP
.B \-mi  \fIMINLUMINANCE\fR, \fB\-\-min-luminance \fIMINLUMINANCE
Minimum luminance (cd/m2) for the perceptual transfer function. Values below 
this will be clipped. This option only applies to the PQ and LOG transfer
functions, as the other available PTFs use lookup tables.

Default value is 0.005.

.TP
.B \-cs  \fICS\fR, \fB\-\-color-space \fICS
The encoder assumes the input to be in RGB. The RGB values are then transformed to
one of the following color spaces for encoding:

  LUV   :  Lu'v'
  RGB   :  Red, green, blue
  YCBCR :  YCbCr, BT.2020 recommendation
  XYZ   :  CIE 1931 XYZ, with D65 reference white

Default is LUV.

.TP
.B \-b  \fIRATE\fR, \fB\-\-bitrate \fIRATE

.TP
.B \-k  \fIINTERVAL\fR, \fB\-\-keyframe-interval \fIINTERVAL
Set fixed interval between keyframes. If the interval is set to 0, the encoder
automatically determines keyframe positions.

Default is 0.

.TP
.B \-l, \fB\-\-lossless
Enable lossless encoding mode.

.TP
.B \-v, \fB\-\-verbose
Enable verbose mode, to display additional information during the encoding.

.SH EXAMPLES
.TP
\fBlumaenc\fR \fB--input\fR hdr_frame_%05d.exr \fB--frames\fR 1:100 \fB--output\fR hdr_video.mkv

Encode EXR frames hdr_frame_00001.exr -- hdr_frame_00100.exr into HDR video hdr_video.mkv.

.TP
\fBlumaenc\fR [...] \fB--encoding-bitdepth\fR 8 \fB--profile\fR 0 \fB--ptf-bitdepth\fR 8 \fB--color-space\fR YCBCR \fB--transfer-function\fR LOG

Encode at 8 bits, in YCbCr, with logarithmic transfer function. This example is 
possible to view in video players that support VP9. However, quantization artifacts 
will be clearly visible when encoding HDR at only 8 bits.

.TP
\fBpfsin\fR frame%04d.hdr \fB--frames\fR 1:2:100 | \fBlumaenc\fR \fB-o\fR video.mkv

Read the frames stored in Radiance (.hdr) format with the index 0001, 0003, 0005, ..., 0099, encode them and save the resulting video in the file video.mkv. Note that pfstools (http://pfstools.sourceforge.net/) is required for piping frames in any HDR format.

.TP
\fBlumaenc\fR [...] \fB--encoding-bitdepth\fR 10 \fB--color-bitdepth\fR 10 \fB--ptf-bitdepth\fR 10 \fB--profile\fR 2 \fB--transfer-function\fR PQ \fB--color-space\fR YCBCR \fB--max-luminance\fR 1000 \fB--min-luminance\fR 0.01

Encodes an HDR video that is possible to upload to e.g. Youtube and Vimeo. The
settings are equivalent to the HDR10 format, and thus properly recognized by 
applications that support HDR10 material encoded using VP9 and stored with the
Matroska container.

.SH "SEE ALSO"
.BR lumadec (1)
.BR lumaplay (1)



ChangeLog

LumaHDRv 0.0.1 <09.11.2015>
    * LumaHDRv first beta release

LumaHDRv 0.0.2 <08.12.2017>
    * Updated libraries
    	- libebml v1.3.5
    	- libmatroska v1.4.8
    	- libvpx v1.6.1

LumaHDRv 0.0.3 <09.12.2017>
    * Fixed meta data and codec settings, 
      so that encoded HDR videos are compatible
      with e.g. Youtube and Vimeo, if encoded 
      using HDR10 specification.

lumaenc.cpp
/**
 * This file is part of the LumaHDRv package.
 * -----------------------------------------------------------------------------
 * Copyright (c) 2015, The LumaHDRv authors.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, 
 *    this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software 
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 * POSSIBILITY OF SUCH DAMAGE.
 * -----------------------------------------------------------------------------
 *
 * \author Gabriel Eilertsen, gabriel.eilertsen@liu.se
 *
 * \date Oct 7 2015
 */

#include <luma_encoder.h>
#include "exr_interface.h"
#include "pfs_interface.h"
#include "luma_exception.h"
#include "arg_parser.h"

#include <iostream>
#include <string.h>

#include "config.h"

// Input and output specific information
struct IOData
{
    IOData() : startFrame(1), endFrame(9999), stepFrame(1), verbose(0)
    {}
    
    std::string hdrFrames, outputFile;
    unsigned int startFrame, endFrame, stepFrame;
    bool verbose;
};

// Determine file extension
bool hasExtension( const char *file_name, const char *extension )
{
    if( file_name == NULL )
        return false;
    size_t fn_len = strlen( file_name );
    size_t ex_len = strlen( extension );

    if( ex_len >= fn_len )
        return false;

    if( strcasecmp( file_name + fn_len - ex_len, extension ) == 0 )
        return true;  

    return false;
}

// Parse a frame range, given as <startFrame>:<step>:<endFrame>
bool getFrameRange(std::string &frames, unsigned int &start, unsigned int &step, unsigned int &end)
{
    int nrDelim = -1;
    std::string::size_type pos = -1;
    
    do
    {
        pos = frames.find_first_of(":", pos+1);
        nrDelim++;
    } while (pos != std::string::npos);

    if (nrDelim < 1 || nrDelim > 2)
        return 0;
    
    unsigned int *range[3] = {&start, &step, &end};
    char *startPtr = &frames[0], *endPtr;
    
    for (size_t i=0; i<3; i+=3-nrDelim)
    {
        *(range[i]) = strtol(startPtr, &endPtr, 10);
        if (startPtr == endPtr)
            return 0;
        startPtr = endPtr + 1;
    }
    
    return 1;
}

// Parse parameter options from command line
bool setParams(int argc, char* argv[], LumaEncoderParams *params, IOData *io)
{
    std::string frames;
    std::string ptf, ptfValues[] = {"PSI", "PQ", "LOG", "HDRVDP", "LINEAR"}; // valid ptf input values
    std::string cs, csValues[] = {"LUV", "RGB", "YCBCR", "XYZ"}; // valid color space input values
    unsigned int bdValues[] = {8, 10, 12}; // valid bit depths

    // Application usage info
    std::string info = std::string("lumaenc -- Compress a sequence of high dyncamic range (HDR) frames in to a Matroska (.mkv) HDR video\n\n") +
                       std::string("Usage: lumaenc --input <hdr_frames> \\\n") +
                       std::string("               --frames <start_frame:step:end_frame> \\\n") +
                       std::string("               --output <output>\n");
    std::string postInfo = std::string("\nExample: lumaenc -i hdr_frame_%05d.exr -f 1:100 -o hdr_video.mkv\n\n") +
                           std::string("See man page for more information.");
    ArgParser argHolder(info, postInfo);

    // Input arguments
    argHolder.add(&io->hdrFrames,            "--input",             "-i",   "Input HDR video sequence");
    argHolder.add(&io->outputFile,           "--output",            "-o",   "Output location of the compressed HDR video", 0);
    argHolder.add(&frames,                   "--frames",            "-f",   "Input frames, formatted as startframe:step:endframe");
    argHolder.add(&params->fps,              "--framerate",         "-fps", "Framerate of video stream, specified as frames/s");
    argHolder.add(&params->profile,          "--profile",           "-p",   "VP9 encoding profile", (unsigned int)(0), (unsigned int)(3));
    argHolder.add(&params->quantizerScale,   "--quantizer-scaling", "-q",   "Scaling of the encoding quantization", (unsigned int)(0), (unsigned int)(63));
    argHolder.add(&params->preScaling,       "--pre-scaling",       "-sc",  "Scaling of pixels to apply before tranformation and encoding", 0.0f, 1e20f);
    argHolder.add(&params->ptfBitDepth,      "--ptf-bitdepth",      "-pb",  "Bit depth of the perceptual transfer function", (unsigned int)(0), (unsigned int)(16));
    argHolder.add(&params->colorBitDepth,    "--color-bitdepth",    "-cb",  "Bit depth of the color channels", (unsigned int)(0), (unsigned int)(16));
    argHolder.add(&ptf,                      "--transfer-function", "-ptf", "The perceptual transfer function used for encoding", ptfValues, 5);
    argHolder.add(&cs,                       "--color-space",       "-cs",  "Color space for encoding", csValues, 4);
    argHolder.add(&params->maxLum,           "--max-luminance",     "-ma",  "Maximum luminance in encoding (for PQ and LOG transfer function)", 100.0f, 1e5f);
    argHolder.add(&params->minLum,           "--min-luminance",     "-mi",  "Minimum luminance in encoding (for PQ and LOG transfer function)", 1e-10f, 99.99f);
    argHolder.add(&params->bitrate,          "--bitrate",           "-b",   "HDR video stream target bandwidth, in Kb/s", (unsigned int)(0), (unsigned int)(9999));
    argHolder.add(&params->keyframeInterval, "--keyframe-interval", "-k",   "Interval between keyframes. 0 for automatic keyframes", (unsigned int)(0), (unsigned int)(9999));
    argHolder.add(&params->bitDepth,         "--encoding-bitdepth", "-eb",  "Encoding at 8, 10 or 12 bits", bdValues, 3);
    argHolder.add(&params->lossLess,         "--lossless",          "-l",   "Enable lossless encoding mode");
    argHolder.add(&io->verbose,              "--verbose",           "-v",   "Verbose mode");

    // Parse arguments
    if (!argHolder.read(argc, argv))
        return 0;
    
    // Check output format
    if (!hasExtension(io->outputFile.c_str(), ".mkv"))
        throw ParserException("Unsupported output format. HDR video should be stored as Matroska file (.mkv)");
    
    // Parse frame range
    if (frames.size() > 0 && !getFrameRange(frames, io->startFrame, io->stepFrame, io->endFrame))
        throw ParserException(std::string("Unable to parse frame range from '" + frames + "'. Valid format is startframe:step:endframe").c_str());
    
    // Valid frame range?
    if (io->endFrame < io->startFrame)
        throw ParserException(std::string("Invalid frame range '" + frames + "'. End frame should be >= start frame").c_str());

    // Translate input strings to enums
    if (!strcmp(ptf.c_str(), ptfValues[0].c_str()))
        params->ptf = LumaQuantizer::PTF_PSI;
    else if (!strcmp(ptf.c_str(), ptfValues[1].c_str()))
        params->ptf = LumaQuantizer::PTF_PQ;
    else if (!strcmp(ptf.c_str(), ptfValues[2].c_str()))
        params->ptf = LumaQuantizer::PTF_LOG;
    else if (!strcmp(ptf.c_str(), ptfValues[3].c_str()))
        params->ptf = LumaQuantizer::PTF_JND_HDRVDP;
    else if (!strcmp(ptf.c_str(), ptfValues[4].c_str()))
        params->ptf = LumaQuantizer::PTF_LINEAR;

    if (!strcmp(cs.c_str(), csValues[0].c_str()))
        params->colorSpace = LumaQuantizer::CS_LUV;
    else if (!strcmp(cs.c_str(), csValues[1].c_str()))
        params->colorSpace = LumaQuantizer::CS_RGB;
    else if (!strcmp(cs.c_str(), csValues[2].c_str()))
        params->colorSpace = LumaQuantizer::CS_YCBCR;
    else if (!strcmp(cs.c_str(), csValues[3].c_str()))
        params->colorSpace = LumaQuantizer::CS_XYZ;

    return 1;
}

int main(int argc, char* argv[])
{
    // Holder for input/output options
    IOData io;
    
    // Encoder and encoder params
    LumaEncoder encoder;
    LumaEncoderParams params = encoder.getParams();

#ifdef HAVE_PFS    
    PfsInterface pfs; // Needs to store state between frames
#endif

    try
    {
        // Read parameters from input
        if (!setParams(argc, argv, &params, &io))
            return 1;
        
        // Set the encoder parameters    
        encoder.setParams(params);

#define STRBUF_LEN 500
        char str[STRBUF_LEN];
        int encoded_frame_count = 0;    
        for (unsigned int f = io.startFrame; f <= io.endFrame; f+=io.stepFrame)
        {        
            LumaFrame frame;

            // Read hdr frames, if available
            if( io.hdrFrames.size() == 0 || hasExtension( io.hdrFrames.c_str(), "pfs" ) )
            {
#ifdef HAVE_PFS
                if( !pfs.readFrame(io.hdrFrames.c_str(), frame) )
                    break;
#else
                throw LumaException( "Compiled without pfstools support" );          
#endif
            }
            else if ( strcmp( io.hdrFrames.c_str(), "__test__" ) == 0 )
                ExrInterface::testFrame(frame);
            else // OpenEXR?          
            {
                snprintf(str, STRBUF_LEN-1, io.hdrFrames.c_str(), f);
                ExrInterface::readFrame(str, frame);
            }

            // Initialize encoder
            if (!encoder.initialized())
                encoder.initialize(io.outputFile.size() == 0 ? "output.mkv" : io.outputFile.c_str(), frame.width, frame.height, io.verbose);

            fprintf(stderr, "Encoding frame %d... ", f);

            // Run the encoder
            encoder.encode(&frame);
            encoded_frame_count++;
            fprintf(stderr, "done\n");
        }

        encoder.finish();
        fprintf(stderr, "\n\nEncoding finished. %d frames encoded.\n", encoded_frame_count);

    }
    catch (ParserException &e)
    {
        fprintf(stderr, "\nlumaenc input error: %s\n", e.what());
        return 1;
    }
    catch (LumaException &e)
    {
        fprintf(stderr, "\nlumaenc encoding error: %s\n", e.what());
        return 1;
    }
    catch (std::exception & e)
    {
        fprintf(stderr, "\nlumaenc error: %s\n", e.what());
        return 1;
    }


    return 0;
}

